src/Symfony/Component/Messenger/Bridge/Amqp/Transport/Connection.php

```diff
@@ -351,41 +359,42 @@ private function getDelayExchange(): \AMQPExchange
 
     /**
      * Creates a delay queue that will delay for a certain amount of time.
-     *
-     * This works by setting message TTL for the delay and pointing
-     * the dead letter exchange to the original exchange. The result
-     * is that after the TTL, the message is sent to the dead-letter-exchange,
-     * which is the original exchange, resulting on it being put back into
-     * the original queue.
      */
-    private function createDelayQueue(int $delay, ?string $routingKey): \AMQPQueue
+    private function createDelayQueue(int $delay, ?string $originQueue, ?string $routingKey): \AMQPQueue
     {
         $queue = $this->amqpFactory->createQueue($this->channel());
-        $queue->setName(str_replace(
-            ['%delay%', '%exchange_name%', '%routing_key%'],
-            [$delay, $this->exchangeOptions['name'], $routingKey ?? ''],
-            $this->connectionOptions['delay']['queue_name_pattern']
-        ));
+
+        if (null === $originQueue) {
+            $dlExchange = $this->exchangeOptions['name'];
+            $dlRoutingKey = $routingKey ?? '';
+            $dlQueueName = $this->getRoutingKeyForDelay($delay, $routingKey);
+        } else {
+            $dlExchange = self::DEFAULT_EXCHANGE;
+            $dlRoutingKey = $originQueue;
```

Comment:
> why this? x-dead-letter-routing-key is the routing key after it being dead lettered. it should be the original routing key. this will set the origin queue as routing key after dead lettering. 

such behavior is required only for delay logic

> so this will change the routing key of the original message which seems buggy.

in case of retry logic it's not buggy. Message has been already routed to the target queue. Original routing key is useless now.  We have all necessary information to route message back to the origin queue. So why to start full publication process again? For me direct routing to the origin queue sounds even more safely and correctly.

> publication with the correct routing key to the default exchange is based on the queue name ($dlQueueName).

i am not sure what you mean here. Could you explain?